### PR TITLE
block already verified email in 'verified_email' route

### PR DESCRIPTION
**Pull request by neverlish**
## Specify project
server side issue

## Description
RESTful route `/verify_email/:email/:hashed` exists for mark user verified. This route works even though user already verified. I changed this route to be blocked when user already verified.

## Related Issues
I didn't created issue. If I need to create, please comment.

## Tests
I didn't created test code. If I need to create, please comment.

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn lint && yarn tsc`
- [ ] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

**Comment by neverlish** [comment]
i will add and commit test case related to changes in pull request

**Comment by LiteHell** [comment]
How about putting the code after validation, not before?

File: server/src/app.ts [hunk 1]
```diff
@@ -81,6 +82,18 @@ export const createApp = (): express.Application => {
...
     try {
+      const user = await prisma.user.findOne({
+        where: {
+          email,
+        },
+      });
+      const alreadyVerified = user && user.verified;
+
+      if (alreadyVerified) {
+        res.send('Already verified email');
```
Comment (hyochan) [inline server/src/app.ts:93]: What about `res.sendStatus(404)`?

Comment (hyochan) [reply server/src/app.ts:93]: Other users & user with the `url` might not need to know if current email is already verified.

**Comment by neverlish** [comment]
> How about putting the code after validation, not before?

- validation you mentioned is `validateCredential` function?
- Can you write comment what is the merit of putting this code after validation? If you do, we can have better discussion.

File: server/src/app.ts (hunk 1 above)
Comment (neverlish) [reply server/src/app.ts]: I changed this code to `res.sendStatus(404)`

**Comment by LiteHell** [comment]
> > How about putting the code after validation, not before?
> 
> - validation you mentioned is `validateCredential` function?
> - Can you write comment what is the merit of putting this code after validation? If you do, we can have better discussion.
> 

- Yes, The validation i mentioned is `validateCredential` function.
- It can be used for email harvesting when it accepts invalid request.
- Also, there is no reason to give any information ("Email already verified" is also information) to invalid request. I think the response should be "invalid request" or "Some error occured!" when request is invalid.

**Comment by neverlish** [comment]
> * Yes, The validation i mentioned is `validateCredential` function.
> * It can be used for email harvesting when it accepts invalid request.
> * Also, there is no reason to give any information ("Email already verified" is also information) to invalid request. I think the response should be "invalid request" or "Some error occured!" when request is invalid.

Thank you for your comment. :)

I changed code lines. https://github.com/dooboolab/hackatalk/pull/186/commits/9741345cb8b2c2e3b4817569f6c871ca9386db99

**Comment by hyochan** [comment]
@neverlish Could you kindly `rebase` this? Then I am going to quickly merge this!

**Comment by neverlish** [comment]
> @neverlish Could you kindly `rebase` this? Then I am going to quickly merge this!

I merged my changed into one commits, rebased onto recent upstream master, and force-pushed.
So it seems like no conflict exists with master branch

File: server/src/app.ts [hunk 2]
```diff
@@ -94,8 +95,21 @@ export const createApp = (): express.Application => {
 
     try {
       const validated = verifyEmailToken(token, req.appSecret);
-      if (validated?.email && validated.type === 'verifyEmail') {
```
Comment (neverlish) [inline server/src/app.ts:97]: I remove optional-chaining because result of `verifyEmailToken` is non-nullable

**Comment by hyochan** [comment]
@all-contributors please add @neverlish for code.

**Pull request merged**